accepted/2020/minimum-os-version/minimum-os-version.md

```diff
@@ -0,0 +1,245 @@
+
+To enable tooling to provide a good experience, the reference assemblies must be annotated with information about the availability of the APIs, as is current done for Xamarin Apple platforms.
+
+The following attributes are proposed for the `System.Runtime.Versioning` namespace. They are based on the Xamarin attributes, but they do not have the `Deprecated` availability kind (it’s redundant), they rename the `Unavailable` availability kind to `Removed` (for consistency), and they do not have the ability to differentiate availability based on architecture (reduction of scope).
+
+```csharp
+[AttributeUsage(AttributeTargets.All, AllowMultiple=true)]
+public abstract class PlatformAvailabilityAttribute
+{
+    internal PlatformAvailabilityAttribute(
+    string platformIdentifier,
+    int majorVersion, int minorVersion, int servicingVersion, int bindingRevision
+    string message
+    ) {}
+    public AvailabilityKind AvailabilityKind { get; }
+    public string Message { get; }
+    public string PlatformIdentifier { get; }
+    public Version OSVersion { get; }
+    public int BindingRevision { get; }
+}
+
+public enum AvailabilityKind
+{
+    Introduced,
+    Obsoleted,
+    Removed
+}
+```
+
+The platform string is expected to have the same values as used in the `TargetPlatformIdentifier` portion of platform-specific TFMs. Having it in the availability attributes may seem redundant as it will also be burned into assembly attributes for platform-specific reference assemblies. However, explicitly specifying the attribute is useful for platform-agnostic `net5.0` reference assemblies that have methods that are only available on some platforms e.g. due to having platform-specific implementations via bait-and-switch. This may also improve source readability when multi-targeting as it will remove the need to use `#if` around the annotations.
+
+Constants for the platform identifiers could perhaps be exposed on a `PlatformIdentifier` type.
+
+The values declared on this attribute are expected to be accessible by extracting constructor arguments from metadata. However, for readability it’s expected to be applied using subclasses with more specific names. To enforce this, it has an internal constructor and its subclasses are sealed.
+
+The concrete subclasses are `IntroducedAttribute`, `ObsoletedAttribute` and `RemovedAttribute` subclasses, which take of the following form:
+
+```csharp
+public sealed class IntroducedAttribute : PlatformAvailabilityAttribute
+{
+    public IntroducedAttribute(
+    string platform,
+    int majorVersion,
+    int minorVersion = 0,
+    int servicingVersion = 0,
+    int bindingRevision = 0,
+    string message = null
+    ) : base(platform, majorVersion, minorVersion, servicingVersion, bindingRevision, message) {}
+}
+```
+
+These attributes are used as follows:
+
+```csharp
+[Introduced(PlatformIdentifier.macOS, 10, 2)]
+[Deprecated(PlatformIdentifier.macOS, 10, 9)]
+[Removed(PlatformIdentifier.macOS, 10, 12, 1)]
```

Comment:
Why not use solutions like `WinMD` and `XLang` to generate metadata (bindings) about the platforms' APIs instead manually duplicating the Swift's attributes into .NET attributes?

That way bindings can be language/framework neutral and in unified format.

Reply:
This is a super simple and inexpensive mechanism. We don't want to take on heavy weight dependencies for it. It's also "OK" to duplicate attributes like this.